whois-api/src/main/java/net/ripe/db/whois/api/rdap/RdapRelationService.java

```diff
@@ -0,0 +1,213 @@
+                .findFirst().orElse(null);
+
+        mostSpecificFillingOverlaps.add(firstResourceIpEntry);
+
+        final IpInterval firstSibling = (IpInterval)siblingsAndExact.getFirst().getKey();
+        final IpInterval lastSibling = (IpInterval)siblingsAndExact.getLast().getKey();
+
+        final IpEntry parent = (IpEntry) ipTree.findFirstLessSpecific(IpInterval.parse(mostSpecificResource.toString())).getFirst();
+        final IpInterval parentInterval = (IpInterval) parent.getKey();
+
+        if (!parentInterval.equals(reverseIp) &&
+                !childrenCoverParentRange(firstSibling, lastSibling, parentInterval)){
+            extractBottomMatches(ipTree, reverseIp, parentInterval, mostSpecificFillingOverlaps);
+        }
+    }
+
+    private static boolean childrenCoverParentRange(final IpInterval firstResource, final IpInterval lastResource, final IpInterval parent){
+        if (firstResource instanceof Ipv4Resource ipv4Resource){
+            final Ipv4Resource lastIpv4Resource = (Ipv4Resource)lastResource;
+            final Ipv4Resource parentIpv4Resource = (Ipv4Resource)parent;
+            return ipv4Resource.begin() == parentIpv4Resource.begin() && lastIpv4Resource.end() == parentIpv4Resource.end();
+        }
+
+        final Ipv6Resource ipv6Resource = (Ipv6Resource)firstResource;
+        final Ipv6Resource lastIpv6Resource = (Ipv6Resource)lastResource;
+        final Ipv6Resource parentIpv6Resource = (Ipv6Resource)parent;
+        return Objects.equals(ipv6Resource.begin(), parentIpv6Resource.begin()) && Objects.equals(lastIpv6Resource.end(), parentIpv6Resource.end());
+    }
+
+    private static List<IpEntry> findSiblingsAndExact(final IpTree ipTree, final IpInterval parentResource) {
+        final List<IpEntry> parentList = ipTree.findFirstLessSpecific(parentResource);
+        if (parentList.isEmpty()){
+            return ipTree.findExact(parentResource);
+        }
+        return ipTree.findFirstMoreSpecific(IpInterval.parse(parentList.getFirst().getKey().toString()));
+    }
+
+    private IpEntry searchFirstLessSpecific(final IpTree ipTree, final IpInterval reverseIp){
+        final List<IpEntry> parentList = ipTree.findFirstLessSpecific(reverseIp);
+        if (parentList.isEmpty() || !resourceExist(reverseIp, parentList.getFirst())){
+            throw new RdapException("404 Not Found", "No up level object has been found for " + reverseIp.toString(), HttpStatus.NOT_FOUND_404);
+        }
+        return parentList.getFirst();
+    }
+
+    private List<IpEntry> searchTopLevelResource(final IpTree ipTree, final IpInterval reverseIp){
+        IpEntry ipEntry;
+        try {
+            ipEntry = searchFirstLessSpecific(ipTree, reverseIp);
+        } catch (RdapException ex){
+            throw new RdapException("404 Not Found", "No top-level object has been found for " + reverseIp.toString(), HttpStatus.NOT_FOUND_404);
+        }
+
+        return List.of(loopUpLevels(ipTree, ipEntry));
```

Comment:
put the return statement inside try block

Reply:
Done